app/api/csv/validateColumns.ts

```diff
@@ -0,0 +1,117 @@
+import { templateUtils } from 'api/templates';
+import _ from 'lodash';
+import { objectIndex } from 'shared/data_utils/objectIndex';
+import { propertyTypes } from 'shared/propertyTypes';
+import { PropertySchema } from 'shared/types/commonTypes';
+import { TemplateSchema } from 'shared/types/templateType';
+import { peekHeaders } from './csv';
+import { notTranslated } from './entityRow';
+import { ImportFile } from './importFile';
+
+const PROPERTIES_WITH_LANGUAGE = new Set([
+  propertyTypes.text,
+  propertyTypes.markdown,
+  propertyTypes.select,
+  propertyTypes.multiselect,
+  propertyTypes.link,
+  propertyTypes.nested,
+  'title',
+]);
+
+class ArrangeColumnsError extends Error {}
+
+const readColumns = async (
+  file: ImportFile,
+  template: TemplateSchema,
+  availableLanguages: string[],
+  newNameGeneration: boolean
+): Promise<{
+  propertiesByName: Record<string, PropertySchema>;
+  headersWithoutLanguage: string[];
+  languagesPerHeader: Record<string, Set<string>>;
+}> => {
+  const propertiesByName = objectIndex(
+    template.properties || [],
+    p => p.name,
+    p => p
+  );
+  propertiesByName.title = { label: 'title', name: 'title', type: 'text' };
+  const headers = await peekHeaders(await file.readStream());
+  const safeNameHeaders = headers.map(h => templateUtils.safeName(h, newNameGeneration));
+  const doesNotHaveLanguage = notTranslated(availableLanguages);
+  const [headersWithoutLanguage, headersWithLanguage] = _.partition(safeNameHeaders, h =>
+    doesNotHaveLanguage(h)
+  );
+  const languagesPerHeader: Record<string, Set<string>> = {};
+  headersWithLanguage.forEach(h => {
+    const [name, l] = h.split('__');
```

Comment:
Is this safe to do? Could it be possible that the name generation leads to something like `word__other__en`?

Reply:
Then, I agree with you and keeping it consistent is the best thing to do now. It may make sense to create a lower priority issue, but also, we haven't received an incident in that regard. So I leave it to you :)